### ISSUE TITLE
如果你一定要TCGA数据库的转录组测序的TPM表达量矩阵，不妨自己进行转换啊！

### ISSUE THREAD

**Issue by ixxmu** [ixxmu/mp_duty#4566]
https://mp.weixin.qq.com/s/kqpoA_Brbbt88GZnc38S2w

**Comment by ixxmu** [comment]
如果你一定要TCGA数据库的转录组测序的TPM表达量矩阵，不妨自己进行转换啊！ by 生信技能树
------
<div><section data-tool="mdnice编辑器" data-website="https://www.mdnice.com"><blockquote data-tool="mdnice编辑器"><p>前面我们的泛癌系列教程，都是直接使用count矩阵进行分析，最多就是进行一些转换而已。但是很多读者不依不饶一定要我们使用TPM表达量矩阵，因为这个更权威或者说更流行！</p></blockquote><p data-tool="mdnice编辑器">首先需要下载TCGA的33种癌症的全部数据，尤其是表达量矩阵和临床表型信息啦，这里我们推荐在ucsc的xena里面下载：https://xenabrowser.net/datapages/，可以看到，确实是没有提供TPM表达量矩阵，但是自己进行转换啊！无论RPKM或FPKM或者TPM格式是多么的遭人诟病，它的真实需求还是存在， 那么我们该如何合理的定义基因的长度呢？</p><p data-tool="mdnice编辑器">这个时候需要理解：<code>基因，转录本(transcripts,isoform，mRNA序列)、EXON区域，cDNA序列、UTR区域，ORF序列、CDS序列</code> 这些概念，一个基因可以转录为多个转录本，真核生物里面每个转录本通常是由一个或者多个EXON组成，能翻译为蛋白的EXON区域是CDS区域，不能翻译的那些EXON的开头和结尾是UTR区域，翻译区域合起来是ORF序列，而转录本逆转录就是cDNA序列。</p><h3 data-tool="mdnice编辑器"><span></span>关于基因长度的探讨<span></span></h3><p data-tool="mdnice编辑器">五年前我分享过前主流定义基因长度的几种方式。</p><ul data-tool="mdnice编辑器"><li><section>挑选基因的<strong>最长转录本</strong></section></li><li><section>选取多个转录本<strong>长度的平均值</strong></section></li><li><section>非冗余外显子(EXON)长度之和</section></li><li><section>非冗余 <strong>CDS</strong>（Coding DNA Sequence） 长度之和</section></li></ul><p data-tool="mdnice编辑器">参考我的博客：定义基因长度为非冗余CDS之和 http://www.bio-info-trainee.com/3298.html</p><p data-tool="mdnice编辑器">其它参考举例：</p><ul data-tool="mdnice编辑器"><li><section>https://www.uniprot.org/uniprot/A2CE44</section></li><li><section>http://www.informatics.jax.org/marker/MGI:3694898</section></li><li><section>http://www.informatics.jax.org/sequence/marker/MGI:3694898?provider=RefSeq</section></li><li><section>https://asia.ensembl.org/Mus_musculus/Gene/Summary?g=ENSMUSG00000073062</section></li><li><section>https://www.ncbi.nlm.nih.gov/CCDS/CcdsBrowse.cgi?REQUEST=CCDS&amp;DATA=CCDS41065.1</section></li></ul><p data-tool="mdnice编辑器">CCDS文件里面：</p><pre data-tool="mdnice编辑器"><span></span><code>X NC_000086.7 Zxdb 668166 CCDS41065.1 Public + 94724674 94727295 [94724674-94727295] Identical<br></code></pre><p data-tool="mdnice编辑器">可以看到不同的定义，得到的基因长度定义还是差异很大的。</p><p data-tool="mdnice编辑器">我这里有一个代码可以获取小鼠的 基因长度，基于 非冗余exon长度之和：</p><pre data-tool="mdnice编辑器"><span></span><code><br><span># BiocManager::install('TxDb.Hsapiens.UCSC.hg38.knownGene')</span><br><span>library</span>(<span>"TxDb.Hsapiens.UCSC.hg38.knownGene"</span>)<br>txdb &lt;- TxDb.Hsapiens.UCSC.hg38.knownGene<br><br><span>## 下面是定义基因长度为 非冗余exon长度之和</span><br><span>if</span>(!file.exists(<span>'gene_length.Rdata'</span>)){<br>  exon_txdb=exons(txdb)<br>  genes_txdb=genes(txdb)<br>  <br>  o = findOverlaps(exon_txdb,genes_txdb)<br>  o<br>  t1=exon_txdb[queryHits(o)]<br>  t2=genes_txdb[subjectHits(o)]<br>  t1=as.data.frame(t1)<br>  t1$geneid=mcols(t2)[,<span>1</span>]<br>  <br>  <span># 如果觉得速度不够，就参考R语言实现并行计算</span><br>  <span># http://www.bio-info-trainee.com/956.html</span><br>  g_l = lapply(split(t1,t1$geneid),<span>function</span>(x){<br>    <span># x=split(t1,t1$geneid)[[1]]</span><br>    head(x)<br>    tmp=apply(x,<span>1</span>,<span>function</span>(y){<br>      y[<span>2</span>]:y[<span>3</span>]<br>    })<br>    length(unique(unlist(tmp)))<br>  })<br>  head(g_l)<br>  g_l=data.frame(gene_id=names(g_l),length=as.numeric(g_l))<br>  <br>  save(g_l,file = <span>'gene_length.Rdata'</span>)<br>} <br><br>load(file = <span>'gene_length.Rdata'</span>)<br>head(g_l)<br><span># 如果是需要其它物种， 换一个包就好了哈！</span><br><span>if</span>(<span>F</span>){<br>  <span># BiocManager::install('TxDb.Mmusculus.UCSC.mm10.knownGene')</span><br>  <span>library</span>(<span>"TxDb.Mmusculus.UCSC.mm10.knownGene"</span>)<br>  txdb &lt;- TxDb.Mmusculus.UCSC.mm10.knownGene<br>  <br>  <br>}<br><br></code></pre><h3 data-tool="mdnice编辑器"><span></span>有了基因长度信息，就很容易转换啦！<span></span></h3><p data-tool="mdnice编辑器">首先检查表达量矩阵额基因长度信息：</p><pre data-tool="mdnice编辑器"><span></span><code>&gt; load(file = <span>'gene_length.Rdata'</span>)<br>&gt; head(g_l)<br>    gene_id length<br>1         1   9488<br>2        10   1285<br>3       100   2809<br>4      1000   9054<br>5     10000  13977<br>6 100009613   1004<br>&gt; load(<span>'../expression/Rdata/TCGA-ACC.htseq_counts.Rdata'</span>)<br><span># 如果你不知道 TCGA-ACC.htseq_counts.Rdata 文件，去看前面的系列教程哦！</span><br>&gt; pd_mat[1:4,1:4]<br>       TCGA-OR-A5JP-01A TCGA-OR-A5JG-01A TCGA-OR-A5K1-01A TCGA-OR-A5JR-01A<br>ZZZ3               1687             1583              784             1555<br>ZZEF1              1955             2036              697             2191<br>ZYX                2528             4567             1924             4639<br>ZYG11B             1541             1225              996             1944<br>&gt; <br></code></pre><p data-tool="mdnice编辑器">可以看到，这个时候一个是基因的id，一个是基因的名字，需要进行继续转换后进行匹配！</p><p data-tool="mdnice编辑器">如果你不知道 TCGA-ACC.htseq_counts.Rdata 文件，去看前面的系列教程哦！前面的泛癌目录是：</p><ul data-tool="mdnice编辑器"><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247506740&amp;idx=3&amp;sn=46e12b5bc3413e52f544c6846481dfb3&amp;scene=21#wechat_redirect" data-linktype="2">estimate的两个打分值本质上就是两个基因集的ssGSEA分析</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247506712&amp;idx=1&amp;sn=2700187268c3af5448eb4c236aa6c684&amp;scene=21#wechat_redirect" data-linktype="2">针对TCGA数据库全部的癌症的表达量矩阵批量运行estimate</a></section></li><li><section><a href="https://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247506712&amp;idx=2&amp;sn=33cbb66d8c21a484a9c5e61924078e4d&amp;scene=21#wechat_redirect" data-linktype="2">不同癌症内部按照estimate的两个打分值高低分组看蛋白编码基因表达量差异</a></section></li></ul><p data-tool="mdnice编辑器">接下来就是核心，转换啦：</p><pre data-tool="mdnice编辑器"><span></span><code><span>library</span>(org.Hs.eg.db)<br>g2s=toTable(org.Hs.egSYMBOL)<br>genelength &lt;-  merge(g_l,g2s,by=<span>'gene_id'</span>)<br>head(genelength)<br>genelength=genelength[,c(<span>3</span>,<span>2</span>)]<br>colnames(genelength) &lt;- c(<span>"gene"</span>,<span>"length"</span>)<br><span>#排序</span><br>counts &lt;- pd_mat[rownames(pd_mat) %<span>in</span>% genelength$gene,]<br><span>#转换成矩阵</span><br>count&lt;- as.matrix(counts)<br><span>#存贮基因长度信息</span><br>genelength[match(rownames(counts),<br>                 genelength$gene),<span>"length"</span>]<br><span># eff_length&lt;- genelength$length</span><br>eff_length &lt;- genelength[match(rownames(counts),genelength$gene),<span>"length"</span>]/<span>1000</span><br>x &lt;- count / eff_length<br>mat_tpm &lt;- t( t(x) / colSums(x) ) * <span>1e6</span> <br>rownames(mat_tpm)&lt;- rownames(count) <br>mat_tpm[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>] <br></code></pre><p data-tool="mdnice编辑器">后续也可以对这样的TPM矩阵，进行一些过滤，一些归一化，可视化比如：</p><p><img data-galleryid="" data-ratio="0.765661252900232" data-s="300,640" data-src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wzCd32a49QiaPtrn0dibLquED7kJR9XapRKXYkhvHDYbRa51GHqfKphRhWXmPsFJibXcTLTjW3qD1RMg/640?wx_fmt=png" data-type="png" data-w="1724" src="https://mmbiz.qpic.cn/mmbiz_png/cZNhZQ6j4wzCd32a49QiaPtrn0dibLquED7kJR9XapRKXYkhvHDYbRa51GHqfKphRhWXmPsFJibXcTLTjW3qD1RMg/640?wx_fmt=png"></p><figure data-tool="mdnice编辑器"><figcaption> </figcaption></figure><p data-tool="mdnice编辑器">代码如下所示：</p><pre data-tool="mdnice编辑器"><span></span><code><br><span># Removing genes that are not expressed</span><br>table(rowSums(mat_tpm==<span>0</span>)== dim(mat_tpm)[<span>2</span>])<br><span># Which values in mat_tpm are greater than 0</span><br>keep.exprs &lt;- mat_tpm &gt; <span>0</span><br><span># This produces a logical matrix with TRUEs and FALSEs</span><br>head(keep.exprs)<br><span># Summary of how many TRUEs there are in each row</span><br>table(rowSums(keep.exprs))<br><span># we would like to keep genes that have at least 2 TRUES in each row of thresh</span><br><span># keep &lt;- rowSums(keep.exprs) &gt;= round(dim(mat_tpm)[2]*0.1)</span><br><span># keep &lt;- rowSums(keep.exprs) &gt;= 0</span><br>keep &lt;- rowSums(keep.exprs) &gt; <span>0</span><br>summary(keep)<br><span># Subset the rows of countdata to keep the more highly expressed genes</span><br>mat_tpm.filtered &lt;- mat_tpm[keep,]<br>rm(keep.exprs)<br>rm(eff_length)<br>mat_tpm.filtered[<span>1</span>:<span>4</span>,<span>1</span>:<span>4</span>]<br><br><span>#--------------------TPM boxplot---------------#</span><br><span>################################################</span><br><span>############## normalize.quantiles #############</span><br><span>################################################</span><br>par(mfrow=c(<span>2</span>,<span>1</span>));<br><span># unnormalised</span><br>boxplot(log2(mat_tpm.filtered+<span>1</span>),xlab=<span>""</span>,ylab=<span>"Log2 transcript per million unnormalised"</span>,las=<span>2</span>,outline = <span>F</span>)<br><span># Let's add a blue horizontal line that corresponds to the median log2TPM</span><br>abline(h=median(log2(mat_tpm.filtered+<span>1</span>)),col=<span>"blue"</span>)<br>title(<span>"Boxplots of log2TPM (unnormalised)"</span>)<br><span># normalised</span><br><span>library</span>(preprocessCore)<br>mat_tpm.norm&lt;- normalize.quantiles(log2(mat_tpm.filtered+<span>1</span>));<br>colnames(mat_tpm.norm)&lt;- colnames(mat_tpm.filtered);<br>rownames(mat_tpm.norm)&lt;- rownames(mat_tpm.filtered);<br>boxplot(mat_tpm.norm,xlab=<span>""</span>,ylab=<span>"Log2 transcript per million normalised"</span>,las=<span>2</span>,outline = <span>F</span>)<br>abline(h=median(mat_tpm.norm),col=<span>"blue"</span>)<br>title(<span>"Boxplots of log2TPM (normalised)"</span>)<br><br>write.table(mat_tpm.filtered,file = <span>"mat_tpm_filtered.txt"</span>,quote = <span>FALSE</span>,sep = <span>"\t"</span>)<br>write.table(mat_tpm.norm,file = <span>"mat_tpm_norm.txt"</span>,quote = <span>FALSE</span>,sep = <span>"\t"</span>)<br>save(mat_tpm.norm,file = <span>"mat_tpm_norm.Rdata"</span>)<br><br><br></code></pre></section><h3 data-tool="mdnice编辑器">文末友情推荐</h3><p data-tool="mdnice编辑器">与十万人一起学生信，你值得拥有下面的学习班：</p><ul data-tool="mdnice编辑器"><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247508762&amp;idx=1&amp;sn=c411bc6007b02e1828235317465fa458&amp;chksm=9b4be3a1ac3c6ab7e649381f828aace86ac811e3edcd152a0a981118d39af4cd87260933b2dc&amp;scene=21#wechat_redirect" textvalue="数据挖掘（GEO,TCGA,单细‍胞）2021第9期" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" wah-hotarea="click">数据挖掘（GEO,TCGA,单细胞）2021第10期</a></section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247508762&amp;idx=2&amp;sn=d1eb0963f32da5ca35a6d873658d0179&amp;chksm=9b4be3a1ac3c6ab78521bec4b27b380f6f560cf8feaabdb744e545e8c7aa2bcfa007fc27dd0d&amp;scene=21#wechat_redirect" textvalue="生信爆款入门-2021第‍9期" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" wah-hotarea="click">生信爆款入门-2021第9期</a></section></li><li><section><a target="_blank" href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&amp;mid=2247508448&amp;idx=1&amp;sn=3283fafda88479411c6fab4e54166fa0&amp;chksm=9b4be15bac3c684d6e9b5c245ce759491c54ab97f39517b266f4d9e9a68f65f2dba56b910a5b&amp;scene=21#wechat_redirect" textvalue="（千呼万唤始出来）生信技能树知识整理实习生在哪里" linktype="text" imgurl="" imgdata="null" data-itemshowtype="0" tab="innerlink" data-linktype="2" wah-hotarea="click">（千呼万唤始出来）生信技能树知识整理实习生在哪里</a></section></li></ul><p><br></p></div>
